File: Go/plane_ticket/service.go
/**
* Authors: Jorge Martins && Diogo Lopes
* This example is from Vasconcelos, V.T. (and several others):
* "Behavioral Types in Programming Languages" (figures 2.4, 2.5 and 2.6)
*/
package main

import(
	"fmt"
	"math/rand"
	"time"
)

func ServiceOrderDelivery(c chan Message){
	fmt.Println("Starting the service!")
	received := <-c //receive address (to store I guess)
	switch received.(type) {
		case CustomerAddress:
			addr := received.(CustomerAddress).customerAddress
			fmt.Printf("Customer Address: Country-> %s, City-> %s, Street-> %s\n", addr.country, addr.city, addr.street)
		default:
			fmt.Println("Error in message type")			
	}

	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	day := r.Intn(30)+1
	month := r.Intn(12)+1
	flightDate := time.Date(2021,time.Month(month),day,0,0,0,0,time.UTC)
	c <- JourneyDate{journeyDate: flightDate}
	fmt.Println("Closing the service!")
}

File: Rust/plane_ticket/src/service.rs
/**
* Authors: Jorge Martins && Diogo Lopes
* This example is from Vasconcelos, V.T. (and several others):
* "Behavioral Types in Programming Languages" (figures 2.4, 2.5 and 2.6)
*/
use crate::message::Message;
use chan::Receiver;
use chan::Sender;
use chrono::prelude::*;
use rand::prelude::*;

pub fn order_delivery(send: Sender<Message>, recv: Receiver<Message>) {
    let service_send = send.clone();
    let service_recv = recv.clone();
    println!("Starting the service!");
    let customer_address_message = service_recv.recv().unwrap();
    match customer_address_message {
        Message::CustomerAddress(addr) => {
            println!(
                "Customer address: Country-> {}, City->{}, Street->{}",
                addr.country(),
                addr.city(),
                addr.street(),
            );
            let mut rng = rand::thread_rng();
            let month: u32 = rng.gen_range(1, 13);
            let day: u32 = rng.gen_range(1, 30);
            let date = Utc.ymd(2021, month, day);
            let flight_date = Message::JourneyDate(date);
            service_send.send(flight_date);
        }
        _ => (),
    }

    println!("Closing the service!");
}
